docs: Changes to the Getting Started document (#103)
FluxML/fluxml.github.io: getting_started.md

```
<br>

### Step 2: Create the training data
First, we'll write a function that generates our "true" data. We'll use to use Flux to recover `W_truth` and `b_truth` by looking only at examples of the `ground_truth` function.

```julia
W_truth = [1 2 3 4 5;
            5 4 3 2 1]

> b_truth = [- 1.0; -2.0]

Reviewer: ```suggestion
b_truth = [-1.0; -2.0]
```

> ground_truth(x) = W_truth * x .+ b_truth

Reviewer: ```suggestion
ground_truth(x) = W_truth*x .+ b_truth
```
To match the function below

Author: Actually, I just totally forgot about those. Thanks for the catch.

```
<br>

Next, we generate our training data by passing random vectors into the ground truth function. We'll also add Gaussian noise using `randn()` so that it's not *too* easy for Flux to figure out the model.

```julia
x_train = [ 5 .* rand(5) for _ in 1:10_000 ]
y_train = [ ground_truth(x) + 0.2 .* randn(2) for x in x_train ]
```

There are two important things to note in this example which differ from real
machine learning problems:
  - Our variables are individual vectors, stored inside another vector. Usually, 
    we would have a collection of N-dimensional arrays (N >= 2) as our data.
  - In a real learning scenario, we would not have access to our ground truth,
    only the training examples.

### Step 3: Define your model

Next, we define the model we want to use to learn the data. We'll use the same form that we used for our training data:

```julia

> model(x) = W*x + b

Reviewer: ```suggestion
model(x) = W*x .+ b
```
To match the function above

```
<br>

### Step 2: Create the training data
First, we'll write a function that generates our "true" data. We'll use to use Flux to recover `W_actual` and `b_actual` by looking only at examples of the `actual_model` function.

```julia
W_actual = [1 2 3 4 5;
            5 4 3 2 1]
b_actual = [- 1.0; -2.0]
actual_model(x) = W_actual * x + b_actual
```
<br>

Next, we generate our training data by passing random vectors into the true model. We'll also add Gaussian noise using `randn()` so that it's not *too* easy for Flux to figure out the model.

```julia
x_train = [ 5 .* rand(5) for _ in 1:10_000 ]

> y_train = [ actual_model(x) + 0.2 .* randn(2) for x in x_train 

Reviewer: ```suggestion
y_train = [ actual_model(x) + 0.2 .* randn(2) for x in x_train]
```
I also think it would help to note what structure `{x,y}_train` actually have. I don't think it's immediately clear that we're dealing with vectors of arrays here, especially since in practice each batch would be a 2D+ array and the outer wrapper "collection" (vector, tuple, dataloader, etc.) is 100% irrelevant.

Author: Would it be preferable to switch this to a 2D array and then use `data = zip(eachcol(x_train), eachcol(y_train))` further down? That was one of my initial thoughts.

Author: > switch this to a 2D array

Nope, this creates awkwardness when generating examples. I'll just clarify this in prose.

```

In most real machine learning scenarios, you would not have access to `actual_model`, and would just get `x_train` and `y_train` as input data.

### Step 3: Define your model

Next, we define the model we want to use to learn the data. We'll use the same form that we used for our training data:

```julia

> training_model(x) = W_learned*x + b_learned

Reviewer: ```suggestion
training_model(x) = W_learned*x .+ b_learned
```
`.+` is good broadcasting hygiene (`+` on arrays calls it behind the scenes). I'd also lean towards calling `training_model` just `model` and renaming `actual_model` to better represent its role as a ground truth function (side note: is "true model" an oxymoron?)

Author: I was suspicious of this since broadcasting is partially what allowed #101 to slip through in the first place (without broadcasting, the dimensions would have failed to match), but you're right in that it's probably better to set a good example here than it is to guard against a one-off case.

Perhaps `model` and `ground_truth`?

Author: Perhaps there is no true model, only a most-useful model.

For this example, we'll define a loss function that measures the squared distance from the predicted output to the actual output:

```julia
function loss(x, y)
  ŷ = training_model(x)
  sum((y - ŷ).^2)
end
```

<br>

### Step 5: Set an optimiser

You train a machine learning model by running an optimization algorithm (optimiser) that finds the best parameters (`W` and `b`). The best parameters for a model are the ones that achieve the best score of the `loss` function. Flux provides [Optimisers](https://fluxml.ai/Flux.jl/stable/training/optimisers/) that you can use to train a model. 

For this tutorial, we'll use a classic gradient descent optimiser with learning rate η = 0.01:

```julia
opt = Descent(0.01)
```
<br>

### Step 6: Train your model

Training a model is the process of computing the gradients with respect to the parameters for each input in the data. At every step, the optimiser updates all of the parameters until it finds a good value for them. This process can be written as a loop: we iterate over the examples in `x_train` and `y_train` and update the model for each example.

> To indicate that we want all derivatives of `W_learned` and `b_learned`, we write `ps = params([W, b])`. This is a convenience function that Flux provides so that we don't have to explicitly list every gradient we want. Check out the section on [Taking Gradients](https://fluxml.ai/Flux.jl/stable/models/basics/#Taking-Gradients) if you want to learn more about how this works.

Reviewer 2: ```suggestion
To indicate that we want all derivatives of `W_learned` and `b_learned`, we write `ps = params(W, b)`. This is a convenience function that Flux provides so that we don't have to explicitly list every gradient we want. Check out the section on [Taking Gradients](https://fluxml.ai/Flux.jl/stable/models/basics/#Taking-Gradients) if you want to learn more about how this works.
```

Author: D'oh. Thanks for the catch.

We can now execute the training procedure for our model:

```julia
train_data = zip(x_train, y_train)
ps = params(W_learned, b_learned)

> for d in train_data

Reviewer: ```suggestion
for (x, y) in train_data
```
and passing them into loss separately, would save us from having to slap down a comment below. I don't recall why the splat was added, but IMO it's wholly unnecessary now.

### Step 4: Define a loss function

A loss function evaluates a machine learning model's performance. In other words, it measures how far the model is from its target prediction. Flux lets you define your own custom loss function, or you can use one of the [Loss Functions](https://fluxml.ai/Flux.jl/stable/training/training/#Loss-Functions-1) that Flux provides. 

For this example, we'll define a loss function that measures the squared distance from the predicted output to the actual output:

```julia
function loss(x, y)
  ŷ = model(x)

>   sum((y - ŷ).^2)

Reviewer: ```suggestion
  sum((y .- ŷ).^2)
```
I understand the want to avoid strange behaviour with broadcasting differently sized operands, but we should be consistent and removing the dot will break broadcast fusion between `-` and `^` (inconsequential here, but a big deal for more expensive forward passes).

Author: Actually, I just totally forgot about those. Thanks for the catch.

<br>

>**Note:** With this pattern, it is easy to add more complex learning routines that make use of control flow, distributed compute, scheduling optimisation, etc. Note that the pattern above is a simple julia *for loop* but it could also be replaced with a *while loop*.

<br>

> Flux lets you do the same process with the [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1) function. `train!` executes a single training step, and you can also put it inside a for-loop to execute multiple training steps. For more information on training a model in Flux, see [Training](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1). 

Reviewer 2: `train!` doesn't do a single training step, it iterates over all the input data, effectively an epoch.

```julia

> for d in train_data

Reviewer 2: See comment above

<br>

>**Note:** With this pattern, it is easy to add more complex learning routines that make use of control flow, distributed compute, scheduling optimisations, etc. Note that the pattern above is a simple Julia *for loop* but it could also be replaced with a *while loop*.

<br>

> While writing your own loop is powerful, sometimes you just want to do the simple thing without writing too much code. Flux lets you do this with the [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1) function, which computes a gradient and updates the model for *every* example in a set (running what is often referred to as a *training epoch*). In our case, we could have replaced the entire loop with the following statement:

Reviewer: ```suggestion
While writing your own loop is powerful, sometimes you just want to do the simple thing without writing too much code. Flux lets you do this with [Flux.train!](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1), which runs one training epoch over a dataset. `Flux.train!` computes gradients and updates model parameters for every sample or batch of samples. In our case, we could have replaced the above loop with the following statement:
```

```julia
Flux.train!(loss, params(W_learned, b_learned), train_data, opt)
```

<br>

For more ways to train a model in Flux, see [Training](https://fluxml.ai/Flux.jl/stable/training/training/#Training-1).

<br>

### Step 7: Examine the Results

The training loop we ran modified `W_learned` and `b_learned` to be closer to the values used to generate the training data (`W_actual` and `b_actual`). We can see how well we did by printing out the difference between the learned and actual matrices.

```julia

> println("The value of W_learned is:")
> display(W_learned)
> print("The largest element-wise difference between W_learned and W_actual is ")
> println(maximum(abs.(W_learned - W_actual))

Reviewer: Instead of writing about what each value represents in prose, I think a simple `@show` would do the trick:
```suggestion
@show W_learned
@show maximum(abs, W_learned .- W_actual)
```

Finally, create a file with extension `.jl` with the code above in any IDE and run it as `julia name-of-your-file.jl `. You can use [Juno IDE](https://junolab.org/) or [Julia VSCode extension](https://www.julia-vscode.org/) to edit and run Julia code. Alternatively, you can run Julia code on a Jupyter notebook (see [IJulia](https://github.com/JuliaLang/IJulia.jl)). Here is the full version of the code:

```julia
using Flux

# Define the ground truth model. We aim to recover W_truth and b_truth using
# only examples of ground_truth()
W_truth = [1 2 3 4 5;
            5 4 3 2 1]
b_truth = [-1.0; -2.0]
ground_truth(x) = W_truth*x .+ b_truth

# Generate the ground truth training data as vectors-of-vectors
x_train = [ 5 .* rand(5) for _ in 1:10_000 ]
y_train = [ ground_truth(x) + 0.2 .* randn(2) for x in x_train ]

# Define and initialize the model we want to train
model(x) = W*x .+ b
W = rand(2, 5)
b = rand(2)

# Define pieces we need to train: loss function, optimiser, examples, and params
function loss(x, y)
  ŷ = model(x)
  sum((y .- ŷ).^2)
end
opt = Descent(0.01)
train_data = zip(x_train, y_train)
ps = params(W, b)

# Execute a training epoch
for (x,y) in train_data

>   gs = Flux.gradient(ps) do

Reviewer: ```suggestion
  gs = gradient(ps) do
```

    loss(x,y)
  end
  Flux.Optimise.update!(opt, ps, gs)
end

# An alternate way to execute a training epoch
# Flux.train!(loss, params(W, b), train_data, opt)

# Print out how well we did
@show W
@show maximum(abs, W .- W_truth)
```

<br>

## What's next

> Congratulations! You have created written and trained a model using Flux. Now, you can continue exploring Flux's capabilities:

Reviewer: ```suggestion
Congratulations! You have created and trained a model using Flux. Now, you can continue exploring Flux's capabilities:
```
